node/external/timemachine/fee/feeComputer.go

```diff
@@ -0,0 +1,110 @@
+package fee
+
+import (
+	"math/big"
+	"sync"
+
+	"github.com/ElrondNetwork/elrond-go-core/data"
+	"github.com/ElrondNetwork/elrond-go/config"
+	"github.com/ElrondNetwork/elrond-go/node/external/timemachine"
+	"github.com/ElrondNetwork/elrond-go/process/economics"
+)
+
+type feeComputer struct {
+	builtInFunctionsCostHandler    economics.BuiltInFunctionsCostHandler
+	economicsConfig                *config.EconomicsConfig
+	penalizedTooMuchGasEnableEpoch uint32
+	gasPriceModifierEnableEpoch    uint32
+	economicsInstances             map[int]economicsDataWithComputeFee
+	mutex                          sync.RWMutex
+}
+
+// NewFeeComputer creates a fee computer which handles historical transactions, as well
+func NewFeeComputer(args ArgsNewFeeComputer) (*feeComputer, error) {
+	err := args.check()
+	if err != nil {
+		return nil, err
+	}
+
+	computer := &feeComputer{
+		builtInFunctionsCostHandler:    args.BuiltInFunctionsCostHandler,
+		economicsConfig:                args.EconomicsConfig,
+		penalizedTooMuchGasEnableEpoch: args.PenalizedTooMuchGasEnableEpoch,
+		gasPriceModifierEnableEpoch:    args.GasPriceModifierEnableEpoch,
+		economicsInstances:             make(map[int]economicsDataWithComputeFee),
+	}
+
+	// Create some economics data instance (but do not save them) in order to validate the arguments:
+	_, err = computer.createEconomicsInstance(0)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = computer.createEconomicsInstance(int(args.GasPriceModifierEnableEpoch))
+	if err != nil {
+		return nil, err
+	}
+
+	return computer, nil
+}
+
+// ComputeTransactionFee computes a transaction fee, at a given epoch
+func (computer *feeComputer) ComputeTransactionFee(tx data.TransactionWithFeeHandler, epoch int) (*big.Int, error) {
+	instance, err := computer.getOrCreateInstance(epoch)
+	if err != nil {
+		return nil, err
+	}
+
+	fee := instance.ComputeTxFee(tx)
+	return fee, nil
+}
+
+// getOrCreateInstance gets or lazily creates a fee computer (using "double-checked locking" pattern)
+func (computer *feeComputer) getOrCreateInstance(epoch int) (economicsDataWithComputeFee, error) {
+	computer.mutex.RLock()
+	instance, ok := computer.economicsInstances[epoch]
+	computer.mutex.RUnlock()
+	if ok {
+		return instance, nil
+	}
+
+	computer.mutex.Lock()
+	defer computer.mutex.Unlock()
+
+	instance, ok = computer.economicsInstances[epoch]
+	if ok {
+		return instance, nil
+	}
+
+	newInstance, err := computer.createEconomicsInstance(epoch)
+	if err != nil {
+		return nil, err
+	}
+
+	computer.economicsInstances[epoch] = newInstance
```

Comment:
can we get OOMs because of using maps that do not get cleared?
Why not using a cacher instead of the map?

Reply:
Using a cache is indeed desirable. Currently, our LRU / time-based caches are non-generic. Though we can work around their functionality and use them in this context, as well, do you think it's OK to postpone until, perhaps, upgrading the go version & support generics (added a TODO).

Howerver, also added a test for analyzing the memory footprint of the fee computer: `~12 MB for 10 000 epochs`.